### PR TITLE
feat: homepage template v2

### LINKED ISSUE #115: Templatise “homepage”
It might be useful to templatise homepage so users can set their own content there (also for us to build SRE Challenges as extensions)

### PR DESCRIPTION
**Pull request by rin-st**
To test: 
Default page:
```
yarn cli
```

Challenge 0 homepage:

```
yarn cli challenge-0-v2-homepage -e scaffold-eth/create-eth-extensions:challenge-0-v2
```

extension (only main page for now): https://github.com/scaffold-eth/create-eth-extensions/tree/challenge-0-v2

Screenshot 2024-09-14 at 15 12 35

---
Pros:
- We have Se-2 title on homepage

Cons:
- We need to move to variable `innerLogic` - logic between start of the component and `return`. If it's not passed, default `const { address: connectedAddress } = useAccount();` will be taken, so if there's no "use client;" or `useAccount` is not imported then page will break; So when there's no `innerLogic`, we need to pass empty string "".

https://github.com/scaffold-eth/create-eth-extensions/blob/caad094bf307f36e8cc952ebf5f948e2b9da9083/extension/packages/nextjs/app/page.tsx.args.mjs#L3

Make extension name dynamic is problematic:
- where to take the name? From branch name? From repo name? In both cases it could be not so fancy
- we need to update templating logic, so it adds complexity

So I think it's much easier to just pass extension name as string if needed

To sum up, I'm tending to v1 for now

Possible solution for avoid `internalLogic`: we can move description with it's inner logic to new component inside SE-2, but someone could want to add the logic to homepage anyway, so again it's not so good

---

Fixes #115

### CODE REVIEW

**Comment by technophile-04** [comment]
Thanks @rin-st !! Just thinking out loud 

Maybe we always show connected Address, and always have "use client" at top? I think its fine?  Maybe we change the placement of Address component by default? This way we remove the alot of complexity. 

Maybe we just allow imports and description. If people want to add some inner logic the can create a new component and and import it in Home page via `imports` args. 

If we don't like Address component not being shown all time maybe then I feel #115 is better idea?

**Comment by rin-st** [comment]
> Maybe we always show connected Address, and always have "use client" at top? I think its fine? Maybe we change the placement of Address component by default? This way we remove the alot of complexity.

Or maybe just remove that address from SE-2/Create-eth homepage? So we won't need to add it to all the extensions

**Comment by technophile-04** [comment]
> Or maybe just remove that address from SE-2/Create-eth homepage? So we won't need to add it to all the extensions

😂 actually that was my first thought before writing that comment! I would personally remove it but I see a little value init. 

What if we remove it and instead add a line of description pointing to component and hooks sections of docs?

**Comment by rin-st** [comment]
Great idea!

You want to make docs links dynamic or not? I mean, just for se-2/create-eth or for + all the extensions? For me both variants looks good. Maybe for all even better

**Comment by technophile-04** [comment]
> You want to make docs links dynamic or not? I mean, just for se-2/create-eth or for + all the extensions? For me both variants looks good. Maybe for all even better

Yeah I think having it always their makes more sense since hooks/components will always be available for each extension. Maybe add it as third card beside Debug and blockexplorer? 

Also cc @carletex before we make this changes, if he agrees with this or not!

**Comment by carletex** [comment]
Adding the connected address was an excuse to add "use client" to the homepage, because of the weird error NextJS was throwing when Austin & people doing workshops (and they tried to use a hook, etc). See https://github.com/scaffold-eth/scaffold-eth-2/pull/721. I think the error has been improved since then.... but I'd personally keep it.

Regarding the docs/links, not sure if I understand. Do we want to add a card/link to the SE-2 docs? Or hooks? 

We can also talk about that in another PR since it won't affect the template, right? Unless we want to make them dynamic.

My idea for this PR was just to do the minimum we need for the SRE Challenges to be an extension. But open to explore more ideas!

**Comment by rin-st** [comment]
> I think the error has been improved since then.... but I'd personally keep it.

Error is still not clear at all. Homepage without `"use client";`.
image

So I think it's better to keep "use client"; on homepage regardless the way we choose.

---

> Do we want to add a card/link to the SE-2 docs? Or hooks?

We're discussing about adding links to both components and hooks. Probably link to main se-2 docs page is enough.

> We can also talk about that in another PR since it won't affect the template, right? Unless we want to make them dynamic.

It won't affect the template if we're just want to remove address. And affect if we want to add link(s) to template right now, so we won't to create multiple pr's to se-2 and create-eth. See options below

---

To be more clear, our options:
1. https://github.com/scaffold-eth/create-eth/pull/119 . Do whatever user wants with homepage
2. This pr. Copying difference from https://github.com/scaffold-eth/create-eth/pull/119 from this pr initial message

Pros:
- We have Se-2 title on homepage

Cons:
- We need to move to variable innerLogic - logic between start of the component and return. If it's not passed, default const { address: connectedAddress } = useAccount(); will be taken, so if there's no "use client;" or useAccount is not imported then page will break; So when there's no innerLogic, we need to pass empty string "".

3. Same as 2, but remove `useAccount` and hence `innerLogic` variable.
4. Same as 3, but add link/links to docs instead of address. Links are static so exists for se-2/create-eth and all the extensions homepages
5. Same as 3, but link/links are dynamic and could be interchanged to extension description

Yes, regarding points 3-5 we need to use "use client" just to avoid errors when people will use hooks there, but actually nothing requires "use client" at the fresh instance. So we need to add the comment somewhere to avoid pr's removing that `use client`, but I think it's not a big problem.

Currently my vision is 3 > 4 > 5 > 1 > 2. Probably 4 would be best but thinking where to add that link(s) so they'd look decent with extensions and can't find a good option
image

**Comment by technophile-04** [comment]
Umm @rin-st I think we should let "use client" and `useAccount`, `Address` demo always be on the page. This way we don't have to go with "innerLogic" route. 

I tried keeping all of them and this is how it looks : 

image


I mean it doesn't look that bad right? Here is the markup for it (we don't need to change anything in base SE-2, used this PR's `.template.mjs` and `.args.mjs` file) 

<details>

<summary>

`page.tsx` : 

</summary>

```tsx
"use client";

import Image from "next/image";
import Link from "next/link";
import type { NextPage } from "next";
import { useAccount } from "wagmi";
import { BugAntIcon, MagnifyingGlassIcon } from "@heroicons/react/24/outline";
import { Address } from "~~/components/scaffold-eth";

const Home: NextPage = () => {
  const { address: connectedAddress } = useAccount();

  return (
    <>
      <div className="flex items-center flex-col flex-grow pt-10">
        <div className="px-5">
          <h1 className="text-center">
            <span className="block text-2xl mb-2">Welcome to</span>
            <span className="block text-4xl font-bold">Scaffold-ETH 2</span>
            {/*Removed the Margin from here in `.template.mjs` file */}
            <span className="block text-xl font-bold">(SpeedRunEthereum extension)</span>
          </h1>
          <div className="flex justify-center items-center space-x-2 flex-col sm:flex-row">
            <p className="my-2 font-medium">Connected Address:</p>
            <Address address={connectedAddress} />
          </div>
          {/* Changed the mt-12 => mt-4 and removed one span from below h1 from `.args.mjs` file */}
          <div className="flex items-center flex-col flex-grow mt-4">
            <div className="px-5 w-[90%]">
              <h1 className="text-center mb-6">
                <span className="block text-4xl font-bold">Challenge #0: Simple NFT</span>
              </h1>
              <div className="flex flex-col items-center justify-center">
                <Image
                  src="/hero.png"
                  width="727"
                  height="231"
                  alt="challenge banner"
                  className="rounded-xl border-4 border-primary"
                />
                <div className="max-w-3xl">
                  <p className="text-center text-lg mt-8">
                    🎫 Create a simple NFT to learn basics of 🏗️ Scaffold-ETH 2. You'll use 👷‍♀️
                    <a
                      href="https://hardhat.org/getting-started/"
                      target="_blank"
                      rel="noreferrer"
                      className="underline"
                    >
                      HardHat
                    </a>{" "}
                    to compile and deploy smart contracts. Then, you'll use a template React app full of important
                    Ethereum components and hooks. Finally, you'll deploy an NFT to a public network to share with
                    friends! 🚀
                  </p>
                  <p className="text-center text-lg">
                    🌟 The final deliverable is an app that lets users purchase and transfer NFTs. Deploy your contracts
                    to a testnet then build and upload your app to a public web server. Submit the url on{" "}
                    <a href="https://speedrunethereum.com/" target="_blank" rel="noreferrer" className="underline">
                      SpeedRunEthereum.com
                    </a>{" "}
                    !
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div className="flex-grow bg-base-300 w-full mt-16 px-8 py-12">
          <div className="flex justify-center items-center gap-12 flex-col sm:flex-row">
            <div className="flex flex-col bg-base-100 px-10 py-10 text-center items-center max-w-xs rounded-3xl">
              <BugAntIcon className="h-8 w-8 fill-secondary" />
              <p>
                Tinker with your smart contract using the{" "}
                <Link href="/debug" passHref className="link">
                  Debug Contracts
                </Link>{" "}
                tab.
              </p>
            </div>
            <div className="flex flex-col bg-base-100 px-10 py-10 text-center items-center max-w-xs rounded-3xl">
              <MagnifyingGlassIcon className="h-8 w-8 fill-secondary" />
              <p>
                Explore your local transactions with the{" "}
                <Link href="/blockexplorer" passHref className="link">
                  Block Explorer
                </Link>{" "}
                tab.
              </p>
            </div>
          </div>
        </div>
      </div>
    </>
  );
};

export default Home;
```

</details>

**Comment by rin-st** [comment]
> Umm @rin-st I think we should let "use client" and useAccount, Address demo always be on the page. This way we don't have to go with "innerLogic" route. And only have "imports" and "description" as args

Yes, you're right! I overcomplicated things a bit and missed this simple case. Will update PR soon, thanks!

**Comment by rin-st** [comment]
Updated Pr and related [extension](https://github.com/scaffold-eth/create-eth-extensions/tree/challenge-0-v2) 

One thing which I don't like is left: "Challenge N" repetition for SRE extensions.
image

But I think it's ok for now and we can change it (or not) when working with extensions.

Yes, we can use only **(SpeedRunEthereum extension)** like in screen here https://github.com/scaffold-eth/create-eth/pull/121#issuecomment-2380415454. But it will be the same for all the extensions, which isn't accurate if being nitpicky since it's different extensions

**Review by technophile-04 — APPROVED** [review]
> One thing which I don't like is left: "Challenge N" repetition for SRE extensions.
> But I think it's ok for now and we can change it (or not) when working with extensions.
> 
> Yes, we can use only **(SpeedRunEthereum extension)** like in screen here [#121 (comment)](https://github.com/scaffold-eth/create-eth/pull/121#issuecomment-2380415454). But it will be the same for all the extensions, which isn't accurate if being nitpicky since it's different extensions

Lol yess, maybe in we can remove Challenge #0 extension from `description` ? 

Screenshot 2024-09-30 at 1 50 29 PM

But I think we can figure this out in `.args.mjs` file while building challenges extension, I think this looks great to me!! Thanks @rin-st!!

File: templates/base/packages/nextjs/app/page.tsx.template.mjs [hunk 1]
```diff
@@ -16,27 +21,13 @@ const Home: NextPage = () => {
           <h1 className="text-center">
             <span className="block text-2xl mb-2">Welcome to</span>
             <span className="block text-4xl font-bold">Scaffold-ETH 2</span>
+            ${externalExtensionName[0] ? `<span className="block text-xl font-bold">(${externalExtensionName[0]} extension)</span>` : ''}
```
[the remark this answers is not in the document]
Comment (rin-st) [reply templates/base/packages/nextjs/app/page.tsx.template.mjs:24]: if we remove it, title will look for example for subgraph

Welcome to
Scaffold-ETH 2
(Subgraph)

for me it's looking absolutely not clear what is `subgraph` here (or ERC20, or onchainkit), so I think to better to have "extension" word

**Comment by rin-st** [comment]
> Lol yess, maybe in we can remove Challenge #0 extension from description

Good point, but not sure again 🤷‍♂️ . I think it should be clear to newcomers that it's challenge #0 and they don't need to search it somewhere on the page

> But I think we can figure this out in .args.mjs file while building challenges extension

Agree! Writing here variants in mind regarding it, to return later
1. Parentheses: (SpeedRunEthereum extension). Title: Challenge #0: Simple NFT
2. Parentheses: (SpeedRunEthereum #0 extension). Title: Challenge #0: Simple NFT
3. Parentheses: (SpeedRunEthereum Challenge #0 extension). Title: Simple NFT
4. Parentheses: (SpeedRunEthereum Challenge #0 extension). Title: Challenge #0: Simple NFT

File: templates/base/packages/nextjs/app/page.tsx.template.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (technophile-04) [reply templates/base/packages/nextjs/app/page.tsx.template.mjs:24]: yup I think lets keep it 🙌

**Comment by carletex** [comment]
Thanks all for this discussion!

> 1. Parentheses: (SpeedRunEthereum extension). Title: Challenge #0: Simple NFT
> 2. Parentheses: (SpeedRunEthereum #0 extension). Title: Challenge #0: Simple NFT
> 3. Parentheses: (SpeedRunEthereum Challenge #0 extension). Title: Simple NFT
> 4. Parentheses: (SpeedRunEthereum Challenge #0 extension). Title: Challenge #0: Simple NFT

I think it's not a big deal and we are too perfeccionist haha. But I have to vote: 2 / 4 >>> 1 / 3. I guess I like verbosity haha. We can also have the "extension" text without bold / smaller.
